### PR TITLE
MM-19508 Extend ShouldProcessMessage to filter out from_webhook posts

### LINKED ISSUE #13259: Extend ShouldProcessMessage to filter out from_webhook posts
Extend [ShouldProcessMessage](https://github.com/mattermost/mattermost-server/blob/a63684fcb5e3ba7b7522b35c29a4cb27779ba823/plugin/helpers_bots.go#L120) to filter out any post with a `from_webhook` prop, unless a new `AllowWebhook` option is true.

Be sure to extend both the [TestProcessMessage](https://github.com/mattermost/mattermost-server/blob/a63684fcb5e3ba7b7522b35c29a4cb27779ba823/plugin/helpers_bots_test.go#L344) unit test to cover this new functionality.

This issue is a follow-up to the changes made for https://mattermost.atlassian.net/browse/MM-16709, and surfaced from the discussion at https://github.com/mattermost/mattermost-server/pull/12539#issuecomment-540782982.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-19508

### PR DESCRIPTION
**Pull request by nadalfederer**
Summary
Extend ShouldProcessMessage to filter out from_webhook posts

Ticket Link
https://mattermost.atlassian.net/browse/MM-19508

closes #13259

### CODE REVIEW

File: plugin/helpers_bots_test.go [hunk 1]
```diff
@@ -447,4 +447,28 @@ func TestShouldProcessMessage(t *testing.T) {
...
 	})
+
+	t.Run("should not process the message which have from_webhook", func(t *testing.T) {
+		channelID := "1"
+		api := setupAPI()
+		api.On("GetChannel", channelID).Return(&model.Channel{Id: channelID, Type: model.CHANNEL_GROUP}, nil)
+		p.API = api
+		api.On("KVGet", plugin.BOT_USER_KEY).Return([]byte(expectedBotId), nil)
+
+		shouldProcessMessage, _ := p.ShouldProcessMessage(&model.Post{ChannelId: channelID, Props: model.StringInterface{"from_webhook": true}}, plugin.AllowBots())
```
Comment (hanzei) [inline plugin/helpers_bots_test.go]: Please also check the returned error

File: plugin/helpers_bots_test.go [hunk 2]
```diff
@@ -447,4 +447,28 @@ func TestShouldProcessMessage(t *testing.T) {
...
+	})
+
+	t.Run("should process the message which have from_webhook with allow webhook plugin", func(t *testing.T) {
+		channelID := "1"
+		api := setupAPI()
+		api.On("GetChannel", channelID).Return(&model.Channel{Id: channelID, Type: model.CHANNEL_GROUP}, nil)
+		p.API = api
+		api.On("KVGet", plugin.BOT_USER_KEY).Return([]byte(expectedBotId), nil)
+
+		shouldProcessMessage, _ := p.ShouldProcessMessage(&model.Post{ChannelId: channelID, Props: model.StringInterface{"from_webhook": true}}, plugin.AllowBots(), plugin.AllowWebhook())
```
Comment (hanzei) [inline plugin/helpers_bots_test.go]: Same here

File: plugin/helpers_bots_test.go [hunk 3]
```diff
@@ -447,4 +447,28 @@ func TestShouldProcessMessage(t *testing.T) {
...
+
+		assert.True(t, shouldProcessMessage)
+	})
```
Comment (hanzei) [inline plugin/helpers_bots_test.go]: Would you also please add a check for a post where `"from_webhook"` is not set and one where `"from_webhook"` is `false`?

File: plugin/helpers_bots.go [hunk 4]
```diff
@@ -157,6 +167,10 @@ func (p *HelpersImpl) ShouldProcessMessage(post *model.Post, options ...ShouldPr
 		return false, nil
 	}
 
+	if post.Props["from_webhook"] == true && !messageProcessOptions.AllowWebhook {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: Could you please move this after the SystemMessage check? Given that we only need to check a field, it's very lightweight and can be done first.

Comment (hanzei) [reply plugin/helpers_bots.go]: Maybe I was ambitious here. I meant moving this code piece to https://github.com/mattermost/mattermost-server/pull/13270/files#diff-7c191d59bfe26e51b1828e40ccfc46f8R150. Does it make my feedback more clear?

File: plugin/helpers_bots.go [hunk 5]
```diff
@@ -138,6 +148,10 @@ func (p *HelpersImpl) ShouldProcessMessage(post *model.Post, options ...ShouldPr
 		return false, nil
 	}
 
+	if !messageProcessOptions.AllowWebhook && post.Props["from_webhook"] == true {
```
Comment (ali-farooq0) [inline plugin/helpers_bots.go]: shouldn't this be `post.Props["from_webhook"] == "true"` instead?

Comment (ali-farooq0) [reply plugin/helpers_bots.go]: or should it be both? because it's a `interface{}` oh boy....

Comment (hanzei) [reply plugin/helpers_bots.go]: Good catch. It should be `post.Props["from_webhook"] == "true"`.

**Comment by hanzei** [comment]
@prapti Gentle reminder to review this. I think the best way to test this is by writing a short example in a plugin. Let me know, if you need help with this.

**Comment by prapti** [comment]
Checked for unit tests and Jira ticket. Will be tested post merge. @hanzei might need some help to setup the test plugin.

**Pull request merged**